crux-calcite/src/crux/calcite.clj

```diff
@@ -0,0 +1,180 @@
+(ns crux.calcite
+  (:require [cheshire.core :as json]
+            [clojure.spec.alpha :as s]
+            [clojure.string :as string]
+            [clojure.tools.logging :as log]
+            [crux.api :as crux]
+            crux.db)
+  (:import java.sql.DriverManager
+           java.util.Properties
+           org.apache.calcite.avatica.jdbc.JdbcMeta
+           [org.apache.calcite.avatica.remote Driver LocalService]
+           [org.apache.calcite.avatica.server HttpServer HttpServer$Builder]
+           [org.apache.calcite.rel.type RelDataTypeFactory RelDataTypeFactory$Builder]
+           [org.apache.calcite.rex RexCall RexInputRef RexLiteral RexNode]
+           org.apache.calcite.sql.SqlKind
+           org.apache.calcite.sql.type.SqlTypeName))
+
+(defonce !crux-nodes (atom {}))
+
+(defprotocol OperandToCruxInput
+  (operand->v [this schema]))
+
+(extend-protocol OperandToCruxInput
+  RexInputRef
+  (operand->v [this schema]
+    (get-in schema [:crux.sql.table/query :find (.getIndex this)]))
+
+  org.apache.calcite.rex.RexCall
+  (operand->v [this schema]
+    (case (str (.-op this))
+      "CRUXID"
+      (keyword (operand->v (first (.-operands this)) schema))
+      (operand->v (first (.-operands this)) schema)))
+
+  RexLiteral
+  (operand->v [this schema]
+    (.getValue2 this)))
+
+(defn- ->operands [schema ^RexCall filter*]
+  (->> (.getOperands filter*)
+       (map #(operand->v % schema))))
+
+(defn- ->crux-where-clauses
+  [schema ^RexNode filter*]
+  (condp = (.getKind filter*)
+    SqlKind/EQUALS
+    [[(apply list '= (->operands schema filter*))]]
+    SqlKind/NOT_EQUALS
+    [(list 'not [(apply list '= (->operands schema filter*))])]
+    SqlKind/AND
+    (mapcat (partial ->crux-where-clauses schema) (.-operands ^RexCall filter*))
+    SqlKind/OR
+    [(apply list 'or (mapcat (partial ->crux-where-clauses schema) (.-operands ^RexCall filter*)))]
+    SqlKind/INPUT_REF
+    [[(list '= (operand->v filter* schema) true)]]
+    SqlKind/NOT
+    [(apply list 'not (mapcat (partial ->crux-where-clauses schema) (.-operands ^RexCall filter*)))]
+    SqlKind/GREATER_THAN
+    [[(list '> (->operands schema filter*))]]
```

Comment:
does this one need to be `apply list` too?

Reply:
Done